### PR TITLE
feature/recaptcha

### PR DESCRIPTION
**Pull request by cochran-at-niche**
This is a library for talking to the reCAPTCHA verification endpoint, for verifying reCAPTCHA v3 tokens generated on the front-end.

For more info, see:
* The [README.md](https://github.com/nicheinc/recaptcha/tree/feature/recaptcha)
* The working [example](https://github.com/nicheinc/recaptcha/tree/feature/recaptcha/example)
* The reCAPTCHA [overview](https://developers.google.com/recaptcha/intro)
* The reCAPTCHA v3 [documentation](https://developers.google.com/recaptcha/docs/v3)

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+// response token is being used within the specified window of time from when
+// the reCAPTCHA was presented. By default, the reCAPTCHA verification endpoint
+// only considers a token valid for 2 minutes, so this method is only necessary
+// if you want to enforce an narrower window than that. Returns
+// *InvalidChallengeTsError if the challenge timestamp is outside the valid
+// window.
+func ChallengeTs(window time.Duration) Criterion {
+	return func(r *Response) error {
+		if diff := now().Sub(r.ChallengeTs); diff > window {
+			return &InvalidChallengeTsError{
```
Comment (juan-carl0s) [inline client.go:219]: This error type could have another field of `time.Duration` type where you could store the `diff` value, that could be a little more friendly when trying to find the duration of time by which the `InvalidChallengeTsError` error was caused.

File: errors.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+// InvalidChallengeTsError is returned from Verify if the ChallengeTs criterion
+// is provided and the response's "challenge_ts" field falls outside the valid
+// window.
+type InvalidChallengeTsError struct {
+	ChallengeTs time.Time
+}
+
+func (e *InvalidChallengeTsError) Error() string {
+	return fmt.Sprintf("invalid reCAPTCHA: invalid challenge timestamp: %s", e.ChallengeTs)
```
Comment (juan-carl0s) [inline errors.go]: I would lean to change this error to communicate that the (token) verification happened too late (since it's being used that way), instead of the timestamp being invalid

File: example/main.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			http.StatusInternalServerError,
+		)
+		return
+	}
+
+	var message string
+	if err := response.Verify(
+		recaptcha.Hostname(strings.Split(*hostnames, ",")...),
+		recaptcha.Action(*action),
+		recaptcha.Score(*score),
```
Comment (juan-carl0s) [inline example/main.go:107]: The `ChallengeTs` `Criterion` could also be added here (since this is the example people would first look at)

File: example/main.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	out, err := json.MarshalIndent(response, "", "    ")
+	if err != nil {
+		http.Error(w,
+			fmt.Sprintf("Error marshalling verification endpoint response: %s\n", err),
+			http.StatusInternalServerError,
+		)
+		return
+	}
+
+	if _, err := w.Write(out); err != nil {
```
Comment (juan-carl0s) [inline example/main.go]: You could remove this whole `w.Write(out)` code block and only write to the ResponseWriter once, farther down in the handler; using `io.WriteString(w, string(out)+message)` instead of `io.WriteString(w, message)`

**Review by juan-carl0s — APPROVED** [review]
This LGTM - I just left a few very small comments
One thing that could save us some lines of code in the future is having a method that in the background just runs `Fetch` and `Verify`, just so when we use this package we could get away with using only one method.
Within the new method and after a successful reCAPTCHA token `Fetch`, the new method could return an error type with a field of type `Response`, so we can still have access to those fields for other purposes (such as logging).

File: example/main.go (hunk 3 above)
Comment (cochran-at-niche) [reply example/main.go:107]: I left it out for the reason described in the ChallengeTs comment:

> By default, the reCAPTCHA verification endpoint only considers a token valid for 2 minutes, so this method is only necessary if you want to enforce an narrower window than that.

In general, I don't think people should need to use that method, and I don't want to encourage people to use it unnecessarily, or confuse them, by including it in the example.

File: errors.go (hunk 2 above)
Comment (cochran-at-niche) [reply errors.go]: I did consider that, but I wanted the errors to be consistent and to clearly indicate which field in the response was the problem (in this case, `challenge_ts`). And I think it's fair to say that the timestamp is "invalid", in the sense that for a challenge timestamp to be valid, it must be within the specified window.

I'm going to add a diff field to the error though, like you mentioned above :+1:

**Comment by cochran-at-niche** [comment]
> One thing that could save us some lines of code in the future is having a method that in the background just runs Fetch and Verify, just so when we use this package we could get away with using only one method.

This is actually one of the things I really disliked about existing reCAPTCHA client libraries - they all bundle the fetching and verification code together into a single method call, which makes it impossible to tell if the token was actually invalid, or if there was just an error trying to hit the endpoint. This could be fixed by wrapping all "fetch" error in one error type, and all "verification" errors in another error type (existing libraries don't do this), and then using a type-switch to differentiate them and handle them differently, but that doesn't actually seem better to me. There's really never a case where you wouldn't want to differentiate those types of errors, so I think it makes more sense to keep the methods separate, so it forces you to think about and handle the errors differently. It also makes the program flow more obvious/apparent (i.e. first you fetch the result from the endpoint, then you verify it).

In other words, I think keeping the methods separate encourages people to use the library the correct way, and discourages people from using it the wrong way (by making it harder to treat all errors the same).
